core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -93,6 +95,38 @@ public DefaultClientRequestContext(
         runThreadLocalContextCustomizer();
     }
 
+    /**
+     * Creates a new derived {@link DefaultClientRequestContext} which only the {@link DefaultRequestLog}
+     * is different from the deriving context. Note that the references of {@link Attribute}s
+     * in the {@link #attrs()} are copied as well.
+     */
+    public DefaultClientRequestContext(DefaultClientRequestContext ctx) {
```

Comment:
How about adding:

- `RequestContext newDerivedContext()` in `RequestContext`
- `@Override ClientRequestContext newDerivedContext()` in `ClientRequestContext`
- `@Override ServiceRequestContext newDerivedContext()` in `ServiceRequestContext`

.. instead of using a constructor?

You could implement `DefaultServiceRequestContext.newDerivedContext()` as well.